### PR TITLE
1100/checksums and malware scanner updates.

### PR DESCRIPTION
**Pull request by tomlovesgithub**
Main branch for work concerning [updates to the virus scanning system](https://docs.google.com/document/d/1k_fgsIQe6h3OBn7JRubmrHYRdA0X2L88/edit).

Includes:
- [x] 1. Checksums
- [x] 2. Upload Status Flag
- [x] 3. File Upload
- [x] 4. File Download
- [x] 5. Admin Tasks

---

### Checksums

Create a new collection in the firestore db to store checksums with the following fields: filepath, checksum, created_at

Write a function which creates a checksum from a file

Create a checksum when a file passes the virus check and store it in the new collection

- pending tests

---

### Upload status flag

Create a new document (/settings/services/fileUpload/enabled) which will store a flag based on the result of a test ran against the virus scanner, which tests that it correctly flags a clean file as clean and an eicar file as infected.

- pending tests

---

### File Upload
covered in [jac-kit 213](https://github.com/jac-uk/jac-kit/pull/213)
---
###  File Download
covered in [jac-kit 213](https://github.com/jac-uk/jac-kit/pull/213)
---
### Admin Tasks
covered in [jac-kit 213](https://github.com/jac-uk/jac-kit/pull/213) and [admin 2536](https://github.com/jac-uk/admin/pull/2536)
---

### CODE REVIEW

File: functions/actions/malware-scanning/scanFile.js [hunk 1]
```diff
@@ -24,40 +23,43 @@ export default (config, firebase) => {
...
         },
-        json: true,
       };
 
-      // make the HTTP request
-      return request(options).then((data) => {
-        return data;
-      }).catch((e) => {
+      const scanResult = await axios(options).catch((e) => {
+        return { error: e.message };
```
Comment (KoWeiJAC) [inline functions/actions/malware-scanning/scanFile.js:48]: Shall we log the e.message of scan API for troubleshooting ?

**Review by KoWeiJAC — APPROVED** [review]

**Review by warrensearle — CHANGES_REQUESTED** [review]

File: functions/actions/malware-scanning/addChecksumToFirestore.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+   * @param {string} filePath - The path of the file.
+   * @param {string} checksum - The checksum to store.
+   * @returns {Promise<string>} - A confirmation message.
+   */
+  async function addChecksumToFirestore(filePath, checksum) {
+    const fileName = path.basename(filePath); // Extract the file name from the file path
+    console.log(`Processing checksum for ${fileName}: ${checksum}`);
+
+    const checksumsCollection = db.collection('checksums');
+    const docRef = checksumsCollection.doc(fileName); // Use the file name as the document ID
```
[the remark this answers is not in the document]
Comment (tomlovesgithub) [reply functions/actions/malware-scanning/addChecksumToFirestore.js:21]: Good point! but files are stored within sub-collections so we shoudn't have this issue hopefully!

**Review by warrensearle — APPROVED** [review]

**Review by drieJAC — APPROVED** [review]

File: functions/actions/malware-scanning/scanFileAndCreateChecksum.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
+import path from 'path';
+import { createChecksum } from './createChecksum.js';
+import { addChecksumToFirestore } from './addChecksumToFirestore.js';
```
Comment (HalcyonJAC) [inline functions/actions/malware-scanning/scanFileAndCreateChecksum.js:3]: We might need to pass parameters when importing the function like:
```js
import addChecksumToFirestoreInit from './addChecksumToFirestore.js';
const { addChecksumToFirestore } = addChecksumToFirestoreInit(firebase, db);
```

File: functions/scheduledFunctions/runScannerTest.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import config from '../shared/config.js';
+import functions from 'firebase-functions';
+import { firebase } from '../shared/admin.js';
+import { runScannerTestFile } from '../actions/malware-scanning/runScannerTest.js';
+const runScannerTest = runScannerTestFile(config, firebase);
```
Comment (HalcyonJAC) [inline functions/scheduledFunctions/runScannerTest.js:5]: We might need to change it to below as `runScannerTest.js` is exported as default and three parameters are required:

```js
import runScannerTestFile from '../actions/malware-scanning/runScannerTest.js';
const runScannerTest = runScannerTestFile(config, firebase, db);
```

File: functions/scheduledFunctions/runScannerTest.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+import config from '../shared/config.js';
+import functions from 'firebase-functions';
+import { firebase } from '../shared/admin.js';
+import runScannerTestInit from '../actions/malware-scanning/runScannerTest.js';
+
+const runScannerTest = runScannerTestInit(config, firebase);
```
Comment (HalcyonJAC) [inline functions/scheduledFunctions/runScannerTest.js:6]: The `db` object needs to be passed into the function.

**Review by HalcyonJAC — APPROVED** [review]
👍

**Pull request merged**